docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,403 @@
+be avoided -- with or without enclave protections.
+
+Enclaves provide new trust boundary protections that address old threats and
+open new capabilities. But just as Open Enclave and TEEs shift the security
+model, application developers must shift how they handle data, especially
+secrets, like keys and passwords.
+<br />
+<br />
+<br />
+
+## 2.1 How _Not_ to Handle Application Secrets
+Let's remind ourselves of a way _not_ to handle application secrets. A common
+example of this well-discussed application weakness is described by
+[CWE-798: Use of Hard-coded Credentials (4.4) (mitre.org)](https://cwe.mitre.org/data/definitions/798.html).
+While application security is often a tradeoff between multiple factors and
+broad edicts are not always appropriate, it's not controversial to simply say
+that hard-coding credentials in application code _should not be done_.
+<br />
+<br />
+<br />
+
+## 2.2 The New, Secure Enclave Way to Handle Application Secrets
+
+Enclaves provide two properties that enable applications to handle secrets
+securely: 1) strong identity, 2) runtime protection of secrets in memory. These
+two properties enable secure point-to-point secret transfers. When applications
+are built with enclaves, all secret management should occur dynamically _inside_
+the enclave and through secure communications with trusted sources, such as
+cloud provider key vaults. Let's look at one example of how enclave attestation
+and enclave memory protection can combine to implement a secure channel.
+
+### Secure Key Release
+One of the core challenges customers have interacting with encrypted
+environments is how to ensure that they can securely and reliably communicate
+with the code running in the environment ("enclave code").
+
+One solution to this problem is what is known as "Secure Key Release", which is
+a pattern that enables this kind of communication with enclave code.
+
+To implement the "Secure Key Release" pattern, the enclave code generates an
+ephemeral asymmetric key. It then serializes the public portion of the key to
+some format (possibly a JSON Web Key, or PEM, or some other serialization
+format).
+The enclave code then calculates a hash (e.g., SHA256) value of the public key
+and passes it as an input to code which generates attestation evidence. For Open
+Enclave, this can be done with `oe_get_evidence`.
+
+The client then sends the evidence and the serialized key to an attestation
+service. The attestation service will appraise the evidence and ensure that the
+hash of the key is present in the evidence and will issue an Attestation Result
+(sometimes called an "Attestation Token").
+
+The client can then send that Attestation Token (which contains the serialized
+key) to a 3rd party "relying party". The relying party then validates that the
+attestation token was created by the attestation service, and thus the
+serialized key can be used to encrypt some data held by the "relying party" to
+send to the service.
```

Comment:
Besides the other problems with this text that I raised, another problem is that the description doesn't match the name.
The name would imply that that a key is only released to an enclave after attestation completes, so it's known to be handed to a trusted client.   In contrast, lines 386-387 instead have the (not yet trusted) enclave _create_ the key and then pass it to the relying party which only _uses_ it if attestation passes.  Thus, it's more like a secure key use than a secure key release.   The only way secure key release would be an accurate name is if you're talking about releasing it to the relying party, but the relying party need not be running in an enclave, and so "secure" doesn't mean the same as what it means in the rest of this article.

Reply:
I agree. Trying to use existing "Secure Key Release" material did not work for this context. I removed the offending material and briefly sketched how attestation could be used with secure data services to manage secrets. It's difficult to go into detail and remain cloud service provider agnostic.

Reply:
Lines 387-404 are "the offending material".  Are you removing those lines in another commit?

Reply:
Addressed in 71be4bcb. I removed design details of the application secrets section. I can see the benefit of a detailed description of how to perform Secure Key Release (for example) in a cloud-agnostic way, but it's beyond the scope of this doc. I would appreciate your review @dthaler.